### PR TITLE
Improve diagnostics messages for invalid custom native type marshalling and cyclical element info.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Use custom resource messages to make the diagnostics more localizable.

Further improvements can come once #1299 is in, since it includes more diagnostics validation for the remaining cases.

### CODE REVIEW

File: DllImportGenerator/DllImportGenerator/Resources.resx [hunk 1]
```diff
@@ -178,6 +184,18 @@
...
+    <value>This element cannot depend on '{0}' for collection size information without creating a dependency cycle</value>
+  </data>
+  <data name="DuplicateCountInfo" xml:space="preserve">
+    <value>Count information for a given element at a given indirection level can only be specified once</value>
+  </data>
+  <data name="DuplicateMarshallingInfo" xml:space="preserve">
+    <value>Multiple marshalling attributes per element per indirection level is unsupported, but duplicate information was provided for indirection level {0}</value>
+  </data>
+  <data name="ExtraneousMarshallingInfo" xml:space="preserve">
+    <value>Marshalling info was specified for 'ElementIndirectionLevel' {0}, but marshalling info was only needed for {1} levels of indirection</value>
```
Comment (elinor-fung) [inline DllImportGenerator/DllImportGenerator/Resources.resx]:
```suggestion
    <value>Marshalling info was specified for 'ElementIndirectionLevel' {0}, but marshalling info was only needed for {1} level(s) of indirection</value>
```

File: DllImportGenerator/DllImportGenerator/Resources.resx [hunk 2]
```diff
@@ -141,6 +141,9 @@
   <data name="ConfigurationNotSupportedMessage" xml:space="preserve">
     <value>The '{0}' configuration is not supported by source-generated P/Invokes. If the specified configuration is required, use a regular `DllImport` instead.</value>
   </data>
+  <data name="ConfigurationNotSupportedMessageMarshallingInfo" xml:space="preserve">
+    <value>The specified marshalling configuration is not supported by source-generated P/Invokes. {0}</value>
```
Comment (elinor-fung) [inline DllImportGenerator/DllImportGenerator/Resources.resx]:
```suggestion
    <value>The specified marshalling configuration is not supported by source-generated P/Invokes. {0}.</value>
```

File: DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs [hunk 3]
```diff
@@ -632,21 +636,26 @@ CountInfo CreateCountInfo(AttributeData marshalUsingData, ImmutableHashSet<strin
 
             if (methods == SupportedMarshallingMethods.None)
             {
-                _diagnostics.ReportConfigurationNotSupported(attrData, "Native Type", nativeType.ToDisplayString());
+                _diagnostics.ReportInvalidMarshallingAttributeInfo(
+                    attrData,
+                    isContiguousCollectionMarshaller ?
```
Comment (elinor-fung) [inline DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs]: nit: I prefer the `?` on the next line (so things following `?`/`:` line up nicely)

**Review by elinor-fung — APPROVED** [review]

File: DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs (hunk 3 above)
Comment (jkoritzinsky) [reply DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs]: I had it like that, but I wasn't sure how we preferred it. I'll change it back.

File: DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs [hunk 4]
```diff
@@ -632,21 +636,26 @@ CountInfo CreateCountInfo(AttributeData marshalUsingData, ImmutableHashSet<strin
...
+                        nameof(Resources.CollectionNativeTypeMustHaveRequiredShapeMessage)
```
Comment (jkoritzinsky) [inline DllImportGenerator/DllImportGenerator/MarshallingAttributeInfo.cs]:
```suggestion
                    isContiguousCollectionMarshaller
                        ? nameof(Resources.CollectionNativeTypeMustHaveRequiredShapeMessage)
```

**Pull request merged**